scalardb/src/scalardb/transfer_append_2pc.clj

```diff
@@ -0,0 +1,263 @@
+                               :tx_id                  :text
+                               :tx_prepared_at         :bigint
+                               :tx_committed_at        :bigint
+                               :tx_state               :int
+                               :tx_version             :int
+                               :before_balance         :int
+                               :before_tx_committed_at :bigint
+                               :before_tx_id           :text
+                               :before_tx_prepared_at  :bigint
+                               :before_tx_state        :int
+                               :before_tx_version      :int
+                               :primary-key            [:account_id :age]})
+
+(defn- prepare-scan
+  [id]
+  (-> (Key. [(IntValue. ACCOUNT_ID id)])
+      (Scan.)
+      (.forNamespace KEYSPACE)
+      (.forTable TABLE)
+      (.withOrdering (Scan$Ordering. AGE Scan$Ordering$Order/DESC))))
+
+(defn- prepare-scan-for-latest
+  [id]
+  (-> id prepare-scan (.withLimit 1)))
+
+(defn- scan-for-latest
+  [tx scan]
+  (first (.scan tx scan)))
+
+(defn- prepare-put
+  [id age balance]
+  (-> (Put. (Key. [(IntValue. ACCOUNT_ID id)]) (Key. [(IntValue. AGE age)]))
+      (.forNamespace KEYSPACE)
+      (.forTable TABLE)
+      (.withValue (IntValue. BALANCE balance))))
+
+(defn- populate-accounts
+  "Insert initial records with transaction.
+  This method assumes that n is small (< 100)"
+  [test n balance]
+  (scalar/retry-when-exception
+   (fn [num]
+     (let [tx (scalar/start-transaction test)]
+       (dotimes [i num]
+         (.put tx (prepare-put i 1 balance)))
+       (.commit tx)))
+   [n]))
+
+(defn- get-balance
+  [^Result result]
+  (-> result (.getValue BALANCE) .get .get))
+
+(defn- get-age
+  [^Result result]
+  (-> result (.getValue AGE) .get .get))
+
+(defn- get-balances
+  [results]
+  (mapv #(get-balance (first %)) results))
+
+(defn- get-ages
+  [results]
+  (mapv #(get-age (first %)) results))
+
+(defn- get-nums
+  [results]
+  (mapv count results))
+
+(defn- calc-new-balance
+  [^Result r amount]
+  (-> r get-balance (+ amount)))
+
+(defn- calc-new-age
+  [^Result r]
+  (-> r get-age inc))
+
+(defn- tx-transfer
+  [tx1 tx2  {:keys [from to amount]}]
+  (try
+    (let [^Result from-result (scan-for-latest tx1 (prepare-scan-for-latest from))
+          ^Result to-result (scan-for-latest tx2 (prepare-scan-for-latest to))]
+      (->> (prepare-put from
+                        (calc-new-age from-result)
+                        (calc-new-balance from-result (- amount)))
+           (.put tx1))
+      (->> (prepare-put to
+                        (calc-new-age to-result)
+                        (calc-new-balance to-result amount))
+           (.put tx2)))
+    (.prepare tx1)
+    (.prepare tx2)
+    (.validate tx1)
+    (.validate tx2)
+    (.commit tx1)
+    (.commit tx2)
+    (catch UnknownTransactionStatusException e
+      (throw e))
```

Comment:
Why not add the tx ID to check the status later?

Reply:
This exception will be caught at the following and add the tx ID there:
https://github.com/scalar-labs/scalar-jepsen/pull/60/files#diff-b44184928fcdbf2b161a9dde329db77cf45587a928546a2142482e8c5247a432R170

Is that okay?